edb/edgeql/tracer.py

```diff
@@ -59,60 +67,85 @@ class Type(NamedObject):
     pass
 
 
-class ObjectType(Type):
+class Source:
+    '''Abstract type that mocks the s_sources.Source for tracing purposes.'''
+    def _init_pointers(self) -> None:
+        self.pointers: Dict[
+            str,
+            Union[s_pointers.Pointer, Pointer]
+        ] = {}
+
+    def getptr(
+        self,
+        schema: s_schema.Schema,
+        name: str,
+    ) -> Optional[Union[s_pointers.Pointer, Pointer]]:
+        return self.pointers.get(name)
+
+
+class ObjectType(Type, Source):
 
-    def __init__(self, name):
+    def __init__(self, name: str) -> None:
         super().__init__(name)
-        self.pointers = {}
+        self._init_pointers()
 
     def is_pointer(self) -> bool:
         return False
 
-    def getptr(self, schema, name):
-        return self.pointers.get(name)
-
 
 class UnionType:
 
-    def __init__(self, types):
+    def __init__(
+        self,
+        types: List[Union[Type, UnionType, so.Object]]
+    ) -> None:
         self.types = types
 
 
-class Pointer:
+class Pointer(Source):
 
-    def __init__(self, name, *, source=None, target=None):
+    def __init__(
+        self,
+        name: str,
+        *,
+        source: Optional[so.Object] = None,
+        target: Optional[s_types.Type] = None,
+    ) -> None:
         self.name = name
         self.source = source
         self.target = target
-        self.pointers = {}
+        self._init_pointers()
 
     def is_pointer(self) -> bool:
         return True
 
-    def getptr(self, schema, name):
-        return self.pointers.get(name)
-
-    def get_target(self, schema):
+    def get_target(
+        self,
+        schema: s_schema.Schema,
+    ) -> Optional[s_types.Type]:
         return self.target
 
-    def get_source(self, schema):
+    def get_source(
+        self,
+        schema: s_schema.Schema,
+    ) -> Optional[so.Object]:
         return self.source
 
-    def get_name(self, schema):
+    def get_name(self, schema: s_schema.Schema) -> str:
         return self.name
 
 
 def trace_refs(
     qltree: qlast.Base,
     *,
-    schema,
-    source: typing.Optional[sn.Name] = None,
-    subject: typing.Optional[sn.Name] = None,
-    path_prefix: typing.Optional[sn.Name] = None,
-    module: typing.Optional[str] = None,
-    objects: typing.Dict[str, object],
-    params: typing.Dict[str, str],
-) -> typing.FrozenSet[sn.Name]:
+    schema: s_schema.Schema,
+    source: Optional[sn.Name] = None,
+    subject: Optional[sn.Name] = None,
+    path_prefix: Optional[Union[str, sn.Name]] = None,
```

Comment:
`Name` is a subclass of `str`, so the Union isn't needed strictly speaking.

Reply:
Hmmm... I'm about 99% certain that mypy complained if I only used `sn.Name`, but let me try removing that to reduce noise, if possible.

Reply:
I've replaced the Union with just `str`, leaving other places where it's pure `sn.Name` unchanged.